source/non_matching/coupling.cc

```diff
@@ -160,23 +176,51 @@ namespace NonMatching
     //        }
     //  }
 
-    for (; cell != endc; ++cell)
+
+    // Get a list of outer cells, qpoints and maps.
+    const auto  cpm = GridTools::compute_point_locations(cache, AllPoints);
+    const auto &all_cells = std::get<0>(cpm);
+    const auto &maps      = std::get<2>(cpm);
+
+    std::vector<
+      std::set<typename Triangulation<dim0, spacedim>::active_cell_iterator>>
+      cell_sets(n_active_c);
+
+    for (unsigned int i = 0; i < maps.size(); ++i)
       {
-        // Reinitialize the cell and the fe_values
-        fe_v.reinit(cell);
-        cell->get_dof_indices(dofs);
+        // Quadrature points should be reasonably clustered:
+        // the following index keeps track of the last id
+        // where the current cell was inserted
+        unsigned int last_id = std::numeric_limits<unsigned int>::max();
+        for (const auto &idx : maps[i])
+          {
+            // Find in which cell the point lies
+            unsigned int cell_id = idx / n_q_points;
+            if (last_id != cell_id)
+              {
+                cell_sets[cell_id].insert(all_cells[i]);
+                last_id = cell_id;
+              }
+          }
+      }
 
-        const std::vector<Point<spacedim>> &Xpoints =
-          fe_v.get_quadrature_points();
+    // Now we run on each cell, get a quadrature formula
+    typename DoFHandler<dim1, spacedim>::active_cell_iterator
+      cell = immersed_dh.begin_active(),
+      endc = immersed_dh.end();
 
-        // Get a list of outer cells, qpoints and maps.
-        const auto  cpm   = GridTools::compute_point_locations(cache, Xpoints);
-        const auto &cells = std::get<0>(cpm);
+    for (unsigned int i = 0; cell != endc; ++cell, ++i)
+      {
+        // Reinitialize the cell
+        cell->get_dof_indices(dofs);
 
-        for (unsigned int c = 0; c < cells.size(); ++c)
+        // List of outer cells
+        const auto &cells = cell_sets[i];
```

Comment:
This is the only place where `i` is used. Hence, you could initialize `i` outside the loop, replace this line by
```
 const auto &cells = cell_sets[i];
```
and switch to a range-based for loop if you prefer that.

Reply:
Ok, I guess you forgot the ++ inside cell_sets, now I have:
`const auto &cells = cell_sets[i++];`
Thank you!

Reply:
Yes, of course.